README.md
ICON SDK for python
========

 ICON supports SDK for 3rd party or user services development.  You can integrate ICON SDK for your project and utilize ICON’s functionality.

 - [Prerequisite](#prerequisite)
 - [Version](#version)
 - [Glossary](#glossary)
 - [Technical information](#technical-information)
 - [Modules](#modules)
 - [Getting started](#getting-started)
 	- [Example](#example)
 - [Functions of wallet](#functions-of-wallet)
 	- [create_keystore_file_of_wallet(keystore_file_path, password)](#createkeystorefileofwallet)
 		- [Arguments](#arguments)
 		- [Successful case](#successful-case)
 		- [Error cases](#error-cases)
 	- [create_wallet_by_private_key(hex_private_key)](#createwalletbyprivatekeyhexprivatekey)
 		- [Arguments](#arguments)
 		- [Successful case](#successful-case)
 		- [Error cases](#error-cases)
 	- [open_keystore_file_of_wallet(keystore_file_path, password)](#openkeystorefileofwalletkeystorefilepath-password)
 		- [Arguments](#arguments)
 		- [Successful case](#successful-case)
 		- [Error cases](#error-cases)
 	- [transfer_value(password, to_address, value, fee=10000000000000000, uri, hex_private_key=None, **kwargs)](#transfervaluepassword-toaddress-value-fee10000000000000000-uri-hexprivatekeynone-kwargs)
 		- [Arguments](#arguments)
 		- [TIP](#tip)
 		- [Successful case](#successful-case)
 		- [Unsuccessful case](#unsuccessful-case)
 		- [Error cases](#error-cases)
 	- [get_wallet_info(uri)](#getwalletinfouri)
 		- [Arguments](#arguments)
 		- [Successful case](#successful-case)
 		- [Error cases](#error-cases)
 	- [get_balance(uri)](#getbalanceuri)
 		- [Arguments](#arguments)
 		- [Successful case](#successful-case)
 		- [Error cases](#error-cases)
 	- [get_address()](#getaddress)
 		- [Arguments](#arguments)
 		- [Successful case](#successful-case)
 		- [Error cases](#error-cases)

# Prerequisite

* Python 3.6.x

# Version

* 0.0.1 beta

# Glossary

* Address of wallet : Unique string to identify the address to transfer value. It begins with hx.

* Private key: A tiny bit of code that is paired with a public key to set off algorithms to encrypt and decrypt a text  for the specific address.

* Public key: Long alphanumeric characters that is used to encrypt data (message).

# Technical information

There are five steps to get from  private->public -> address:

1. Generate a private key.

2. Derive a public key from the private key.

3. H1 = sha3_256( Public key)  => 32 byte

4. BitAddress = last 20 bytes of H1

5. Address = hx || HexString(BitAddress)
ex) hxaa688d74eb5f98b577883ca203535d2aa4f0838c

# Modules

* ```wallet``` : Package name of ICX wallet functions.

# Getting started

```shell
$ pip install iconsdk
```

## Example
```python

from icx.wallet.wallet import Wallet

# Create a keystore file of a wallet.
my_wallet_1, _ = Wallet.create_keystore_file_of_wallet(keystore_file_path="./keystore.txt", password="test1234*")

# Create a wallet by the private key.
my_wallet_2, _ = Wallet.create_wallet_by_private_key(hex_private_key="")

# Open the keystore file of the wallet.
my_wallet_3, _ = Wallet.open_keystore_file_of_wallet(keystore_file_path="./test_keystore_for_transfer.txt", password="ejfnvm1234*")

# Get balance.
balance = my_wallet_1.get_balance(uri="https://testwallet.icon.foundation/api/")

# Get information of the wallet.
wallet_info = my_wallet_1.get_wallet_info(uri="https://testwallet.icon.foundation/api/")

# Get an address.
wallet_address = my_wallet_1.get_address()

# Transfer value 1,010,000,000,000,000,000 loop (1.01 icx) with 10,000,000,000,000,000 loop (0.01 icx) fee.
try:
    result = my_wallet_3.transfer_value(password="ejfnvm1234*", to_address="hx68bc6f60ea01bc033504a217631c601386be26b7", \
                value="1010000000000000000", fee=10000000000000000)
except PasswordIsNotAcceptable:
    print(f"Password is not acceptable.")
except PasswordIsWrong:
     print(f"Password is wrong.")
except AddressIsWrong:
     print(f"Wallet address is wrong.")
except NotEnoughBalanceInWallet:
     print(f"Balance is not enough.")
except TransferFeeIsInvalid:
     print(f"Transaction Fee is invalid. The fee should be 10000000000000000.")
except FeeIsBiggerThanAmount:
     print(f"Fee is bigger than transaction amount.")
except AmountIsInvalid:
     print(f"The amount you want to transfer is not valid.")
except AddressIsSame:
     print(f"Wallet address to transfer must be different from Wallet address to deposit.")

```

# Functions of wallet

## ```create_keystore_file_of_wallet(keystore_file_path, password)```

create both a wallet and a keystore file with file path and given password.

### Arguments

* ```keystore_file_path``` : File path for the keystore file of the wallet.

* ```password```: Password for the wallet. Password must include alphabet character, number, and  special character.

### Successful case

* Return : Instance of Wallet, private key

### Error cases

It will raise following exception.

* ```PasswordIsNotAcceptable```: Password is not acceptable. It must be more than eight characters long, contain any letters from **a** to **z**, any numbers from **0** to **9** and some special characters, including @ (at sign), .(period), -(hyphen or dash), and(or) _ (underscore).

* ```FilePathIsWrong```: File path is wrong.

## ```create_wallet_by_private_key(hex_private_key)```

create wallet without keystore file.

### Arguments

* ```hex_private_key``` : A private key in hexadecimal - 256 bits in hexadecimal is 32 bytes, or 64 characters in the range 0-9 or A-F. A tiny bit of code that is paired with a public key to set off algorithms to encrypt and decrypt a text for the specific address.

### Successful case

* Return : Instance of Wallet, private key

### Error cases

It will raise following exception.

* ```TypeError```

## ```open_keystore_file_of_wallet(keystore_file_path, password)```

Open the created keystore file and read the information of the file.

### Arguments

* ```keystore_file_path``` : File path for the keystore file of the wallet.

* ```password```: Password for the wallet. Password must include alphabet character, number, and  special character.

### Successful case

* Return :  Instance of Wallet.

### Error cases

It will raise following exception.

* ```PasswordIsWrong```: Password is wrong.

* ```FilePathIsWrong```: File path is wrong.

## ```transfer_value(password, to_address, value, fee=10000000000000000, uri, hex_private_key=None, **kwargs)```

Transfer the value from the given wallet to the specific address with the fee.

### Arguments

* ```password``` : Password for the wallet in keystore file used in open_wallet_from_file()

* ```to_address```: Address of the wallet

* ```value``` : Amount of money

* ```fee``` : Transfer fee (10000000000000000 loop)

* ```uri``` : URI of ICON API. The default value is 'https://testwallet.icon.foundation/api/', test net. You can use another URI of ICON API for various test net like Ethereum.

* ```kwargs``` : (Optional) Reserved for the next version

### TIP

* value and fee are integer with decimal point 10^18. Ex) 1.10 icx => 1.10 X 1,000,000,000,000,000,000 = 1,100,000,000,000,000,000 loop.

### Successful case

* Return : Response

``` json
{
    "jsonrpc": "2.0",
    "result": {
        "response_code": 0,
        "tx_hash": "4bf74e6aeeb43bde5dc8d5b62537a33ac8eb7605ebbdb51b015c1881b45b3aed"
    },
    "id":2
}
```
* ```response_code```: JSON RPC error code.
* ```tx_hash```: Hash data of the result. Use icx_getTransactionResult to get the result.
* ```id```: It MUST be the same as the value of the id member in the Request Object.

    * If there was an error in detecting the id in the Request object (e.g. Parse error/Invalid Request), it MUST be Null.

### Unsuccessful case

* Return : Response

```json
{
    "jsonrpc": "2.0",
    "result": {
        "message": "create tx message",
        "response_code": -11
    },
    "id": 2
}
```

### Error cases

It will raise following exception.

* ```AddressIsWrong``` : Wallet address is wrong.

* ```PasswordIsWrong```: Password is wrong.

* ```NoEnoughBalanceInWallet``` : Sender’s wallet does not have enough balance.

* ```TransferFeeIsInvalid``` :  Transfer fee is invalid.

* ```TimestampIsNotCorrect``` : Timestamp is not correct. (Adjust your computer’s time and date.)


## ```get_wallet_info(uri)```

Get the keystore file information and the balance.

### Arguments

* ```uri``` : URI of ICON API. The default value is 'https://testwallet.icon.foundation/api/', test net. You can use another URI of ICON API for various test net like Ethereum.

### Successful case

Return dictionary with sub items like below.

* ```balance``` : the balance of this wallet

* ```depositAddress```: the address of this wallet

* ```completedTransactions``` : list of dictionary to store completed transactions

    * ```requestedTime``` : The time when the transaction was requested

    * ```transactionType``` : deposit or withdraw

    * ```transactionID``` : ID of transaction to track

    * ```amount``` : The amount of money

* ```pendingTransactions``` :  list of dictionary to store pending transactions

    * ```requestedTime``` : The time when the transaction was requested

    * ```transactionType``` : deposit or withdraw

    * ```transactionID``` : ID of transaction to track

    * ```amount``` :  The amount of money

### Error cases

It will raise following exception.

* ```AddressIsWrong``` : Address is wrong.

## ```get_balance(uri)```

Get the balance of all addresses in the current wallet.

### Arguments

* ```uri``` : URI of ICON API. The default value is 'https://testwallet.icon.foundation/api/', test net. You can use another URI of ICON API for various test net like Ethereum.

### Successful case

* Return integer with decimal point 10^18.
Ex) 1.10 icx => It will return 1,100,000,000,000,000,000.

### Error cases

It will raise following exception.

* ```AddressIsWrong``` : Address is wrong.

## ```get_address()```

Get the address of wallet.

### Arguments

* N/A

### Successful case

* Return string of wallet address begins from ‘hx’.

### Error cases

It will raise following exception.

* ```AddressIsWrong``` : Address is wrong.


CHANGELOG.md
# Changelog


## [0.0.6] - July 23, 2018
### Added
N/A

### Changed
 - Fix issue: fixed typo for issue: get_balance_after_transfer().
              changed -- get_balance_after_trasfer() to --> get_balance_after_transfer()
 - Increased time for sleep on an unittest from 1 to 2 because it takes time to make consensus among nodes.

### Removed
N/A


## [0.0.5] - June 18, 2018
### Added
 - get block by hash
 - get block by height
 - get last block
 
### Changed
 - Fix issue: create_wallet_by_private_key() returned the wallet object without the data.

### Removed
N/A


## [0.0.4] - June 12, 2018
### Added
N/A

### Changed
 - Changed import rule

### Removed
N/A


## [0.0.3] - April 27, 2018
### Added
N/A

### Changed
 - Convert README.md to README.rst

### Removed
N/A


## [0.0.2] - April 27, 2018
### Added
N/A

### Changed
 - Rename icon_sdk to iconsdk.

### Removed
N/A


## [0.0.1] - April 26, 2018
### Added
 - Initialize version.

### Changed
N/A

### Removed
N/A


icx/utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2018 theloop Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import base64, hashlib, re, time, os, codecs, json, certifi, requests
from eth_keyfile import create_keyfile_json, extract_key_from_keyfile, load_keyfile
from json import JSONDecodeError
from icx.custom_error import NotEnoughBalanceInWallet, AmountIsInvalid, AddressIsWrong, TransferFeeIsInvalid, \
    FeeIsBiggerThanAmount, NotAKeyStoreFile, AddressIsSame
from icx.signer import IcxSigner


def validate_password(password) -> bool:
    """ Verify the entered password.

    :param password: The password the user entered. type(str)

    :return: bool
    True: When the password is valid format
    False: When the password is invalid format
    """
    return bool(re.match(r'^(?=.*\d)(?=.*[a-zA-Z])(?=.*[!@#$%^&*()_+{}:<>?]).{8,}$', password))


def hex_to_bytes(value):
    return bytes.fromhex(value)


def bytes_to_hex(value):
    return value.hex()


def get_timestamp_us():
    """Get epoch time in us.
    """
    return int(time.time() * 10 ** 6)


def validate_address(address) -> bool:
    try:
        if len(address) == 42 and address.startswith('hx'):
            return True
        raise AddressIsWrong
    except ValueError:
        raise AddressIsWrong


def validate_address_is_not_same(to_address, from_address) -> bool:
    if to_address != from_address:
        return True
    else:
        raise AddressIsSame


def validate_key_store_file(key_store_file_path: object) -> bool:
    """ Check key_store file was saved in the correct format.

    :return: bool
    True: When the key_store_file was saved in valid format.
    False: When the key_store_file was saved in invalid format.
    """
    # The key values ​​that should be in the root location.
    root_keys = ["version", "id", "address", "crypto"]
    crypto_keys = ["ciphertext", "cipherparams", "cipher", "kdf", "kdfparams", "mac"]
    crypto_cipherparams_keys = ["iv"]
    crypto_kdfparams_keys = ["dklen", "salt", "c", "prf"]

    try:
        with open(key_store_file_path, 'rb') as key_store_file:
            key_file = load_keyfile(key_store_file)
        is_valid = has_keys(key_file, root_keys) and has_keys(key_file["crypto"], crypto_keys) and has_keys(key_file["crypto"]["cipherparams"], crypto_cipherparams_keys) and has_keys(key_file["crypto"]["kdfparams"], crypto_kdfparams_keys)
    except KeyError:
        raise NotAKeyStoreFile
    except JSONDecodeError:
        raise NotAKeyStoreFile
    if is_valid is not True:
        raise NotAKeyStoreFile
    return is_valid


def validate_wallet_info(wallet_info: dict) -> bool:
    """ Check a wallet info has the right format or not.

    :param wallet_info:
    :return: bool
    True: When the wallet info is in the correct format.
    False: When the wallet info is in the incorrect format.
    """
    root_keys = ["version", "id", "address", "crypto", "balance"]
    crypto_keys = ["ciphertext", "cipherparams", "cipher", "kdf", "kdfparams", "mac"]
    crypto_cipherparams_keys = ["iv"]
    crypto_kdfparams_keys = ["dklen", "salt", "c", "prf"]

    is_valid = has_keys(wallet_info, root_keys) and has_keys(wallet_info["crypto"], crypto_keys) and \
               has_keys(wallet_info["crypto"]["cipherparams"], crypto_cipherparams_keys) and \
               has_keys(wallet_info["crypto"]["kdfparams"], crypto_kdfparams_keys)
    return is_valid


def validate_block(response_json: dict) -> bool:
    """ Check the response json of the block has right format.

    :param response_json
    :return: bool
    """
    inner_key_name_of_block = ["version", "prev_block_hash","merkle_tree_root_hash","time_stamp","confirmed_transaction_list"]
    is_valid = response_json["result"]["response_code"] is 0 and \
               has_keys(response_json["result"]["block"], inner_key_name_of_block)
    return is_valid


def validate_last_block(response_json: dict) -> bool:
    """ Check the response json of the last block has right format.

    :param response_json
    :return: bool
    """
    inner_key_name_of_result = ["response_code", "block"]
    inner_key_name_of_block = ["version", "prev_block_hash", "merkle_tree_root_hash", "time_stamp", \
                               "confirmed_transaction_list", "block_hash", "height", "peer_id", "signature"]

    is_valid = response_json["result"]["response_code"] is 0 and \
               has_keys(response_json["result"], inner_key_name_of_result) and \
               has_keys(response_json["result"]["block"], inner_key_name_of_block)
    return is_valid


def has_keys(dict_data, key_array):
    for key in key_array:
        if key in dict_data.keys():
            pass
        else:
            return False
    return True


def sha3_256(data):
    """ Get hash value using sha3_256 hash function.

    :param data:
    :return: 256bit hash value (32 bytes). type(bytes)
    """
    return hashlib.sha3_256(data).digest()


def get_address_by_privkey(privkey_bytes):
    """ Get address by Private key.

    :param privkey_bytes: Private key. type(string)
    """
    account = IcxSigner.from_bytes(privkey_bytes)
    return f'hx{bytes_to_hex(account.address)}'


def get_tx_hash(method, params):
    """ Create tx_hash from params object.

    :param method: Method name. type(str)
    :param params: The value of 'params' key in jsonrpc.

    :return: bytes: sha3_256 hash value
    """

    tx_phrase = get_tx_phrase(method, params)
    return sha3_256(tx_phrase.encode())


def get_tx_phrase(method, params):
    """ Create tx phrase from method and params. tx_phrase means input text to create tx_hash.

    :param method: The value of 'params' key in jsonrpc. type(dict)
    :param params: Method name. type(str)

    :return: sha3_256 hash format without '0x' prefix
    """
    keys = [key for key in params]

    key_count = len(keys)
    if key_count == 0:
        return method

    phrase = get_params_phrase(params)

    return f'{method}.{phrase}'


def get_params_phrase(params):
    """Create params phrase recursively
    """
    keys = [key for key in params]
    keys.sort()
    key_count = len(keys)
    if key_count == 0:
        return ""
    phrase = ""

    if isinstance(params[keys[0]], dict) is not True:
        phrase += f'{keys[0]}.{params[keys[0]]}'
    elif bool(params[keys[0]]) is not True:
        phrase += f'{keys[0]}'
    else:
        phrase += f'{keys[0]}.{get_params_phrase(params[keys[0]])}'

    for i in range(1, key_count):
        key = keys[i]

        if isinstance(params[key], dict) is not True:
            phrase += f'.{key}.{params[key]}'
        elif bool(params[key]) is not True:
            phrase += f'.{key}'
        else:
            phrase += f'.{key}.{get_params_phrase(params[key])}'

    return phrase


def sign_recoverable(private_key_bytes, tx_hash_bytes):
    """
    :param private_key_bytes: Byte private key value.
    :param tx_hash_bytes: 32 byte tx_hash data. type(bytes)
    :return: signature_bytes + recovery_id(1)
    """
    signer = IcxSigner.from_bytes(private_key_bytes)
    signature_bytes, recovery_id = signer.sign_recoverable(tx_hash_bytes)

    # append recover_id(1 byte) to signature_bytes.
    return bytes(bytearray(signature_bytes) + recovery_id.to_bytes(1, 'big'))


def sign(private_key_bytes, tx_hash_bytes):
    """
    :param private_key_bytes:
    :param tx_hash_bytes:
    :return: base64-encoded string of recoverable signature data
    """
    recoverable_sig_bytes = sign_recoverable(private_key_bytes, tx_hash_bytes)
    return base64.b64encode(recoverable_sig_bytes)


def create_jsonrpc_request_content(_id, method, params):

    content = {
        'jsonrpc': '2.0',
        'method': method,
        'id': _id
    }

    if params is not None:
        content['params'] = params

    return content


def post(url, payload):
    try:
        path = certifi.where()
        r = requests.post(url, json=payload, verify=path)
        return r
    except requests.exceptions.Timeout:
        raise RuntimeError("Timeout happened. Check your internet connection status.")


def get_payload_of_json_rpc_get_balance(address, url):
    method = 'icx_getBalance'
    params = {'address': address}
    payload = create_jsonrpc_request_content(0, method, params)
    return payload


def check_balance_enough(balance, amount, fee):
    """ Check if the user has enough balance to transfer.

    :param balance: Balance of the user's wallet.
    :param amount: Amount of money. type(str)
    :param fee: Transfer fee.

    :return: True when the user has enough balance.
    """
    if balance >= amount + fee:
        return True
    else:
        raise NotEnoughBalanceInWallet
    pass


def check_amount_and_fee_is_valid(amount, fee):
    if amount <= 0:
        raise AmountIsInvalid
    if fee <= 0 or fee != 10000000000000000:
        raise TransferFeeIsInvalid
    if amount < fee:
        raise FeeIsBiggerThanAmount


def change_hex_balance_to_decimal_balance(hex_balance, place=18):
    """ Change hex balance to decimal decimal icx balance.

    :param: hex_balance

    :return: result_decimal_icx: string decimal icx
    """
    dec_balance = int(hex_balance, 16)
    str_dec_balance = str(dec_balance)
    if dec_balance >= 10 ** place:
        str_int = str_dec_balance[:len(str_dec_balance) - place]
        str_decimal = str_dec_balance[len(str_dec_balance) - place:]
        result_decimal_icx = f'{str_int}.{str_decimal}'
        return result_decimal_icx

    else:
        zero = "0."
        val_point = len(str_dec_balance)
        point_difference = place - val_point
        str_zero = "0" * point_difference
        result_decimal_icx = f'{zero}{str_zero}{dec_balance}'
        return result_decimal_icx


def request_generator(url):
    while True:
        payload = yield
        yield post(url, payload)


def make_params(user_address, to, amount, fee, method, private_key_bytes):
    """ Make params for jsonrpc format.

    :param user_address: Address of user's wallet.
    :param to: Address of wallet to receive the asset.
    :param amount: Amount of money.
    :param fee: Transaction fee.
    :param method: Method type. type(str)
    :param private_key_bytes: Private key of user's wallet.

    :return: type(dict)
    """
    params = {
        'from': user_address,
        'to': to,
        'value': hex(amount),
        'fee': hex(fee),
        'timestamp': str(get_timestamp_us())
    }
    tx_hash_bytes = get_tx_hash(method, params)
    signature_bytes = sign(private_key_bytes, tx_hash_bytes)
    params['tx_hash'] = tx_hash_bytes.hex()
    params['signature'] = signature_bytes.decode()

    return params


def store_wallet(file_path, json_string):
    """ Store wallet information file in JSON format.

    :param file_path: The path where the file will be saved. type: str
    :param json_string: Contents of key_store_file
    """
    if os.path.isfile(file_path):
        raise FileExistsError

    with open(file_path, 'wt') as f:
        f.write(json_string)


def make_key_store_content(password):
    """ Make a content of key_store.

    :param password: Password including alphabet character, number, and special character.

    :return: key_store_content(dict)
    """
    signer = IcxSigner()
    private_key = signer.private_key
    key_store_contents = create_keyfile_json(private_key, bytes(password, 'utf-8'), iterations=262144)
    icx_address = "hx" + signer.address.hex()
    key_store_contents['address'] = icx_address
    key_store_contents['coinType'] = 'icx'
    return key_store_contents


def key_from_key_store(file_path, password):
    """

    :param file_path:
    :return:
    """
    with open(file_path, 'rb') as file:
        private_key = extract_key_from_keyfile(file, password)
    return private_key


def get_balance(address, url):
    """ Get balance of the address indicated by address.

    :param address: icx account address starting with 'hx'
    :param url: api target url

    :return: icx
    """
    url = f'{url}v2'

    method = 'icx_getBalance'
    params = {'address': address}
    payload = create_jsonrpc_request_content(0, method, params)
    response = post(url, payload)
    content = response.json()
    hex_balance = content['result']['response']
    dec_loop_balance = int(hex_balance, 16)

    return dec_loop_balance


def get_block_by_hash(hash, url):
    """ Get block information by hash.

    :param hash: Using hash values ​​with electronic signatures. 64 character. hexadecimal.
    :param url: api target url

    :return: response result(json)
    """
    url = f'{url}v2'

    method = 'icx_getBlockByHash'
    params = {'hash': hash}
    payload = create_jsonrpc_request_content(0, method, params)
    response = post(url, payload)
    json_response = response.json()
    return json_response


def get_block_by_height(height, url):
    """ Get block information by height.

    :param height: block's height
    :param url: api target url

    :return: response result(json)
    """
    url = f'{url}v2'

    method = 'icx_getBlockByHeight'
    params = {'height': height}
    payload = create_jsonrpc_request_content(0, method, params)
    response = post(url, payload)
    json_response = response.json()
    return json_response


def get_last_block(url):
    """ Get last block information.

    :param url: api target url

    :return: response result(json)
    """
    url = f'{url}v2'

    method = 'icx_getLastBlock'
    params = {}
    payload = create_jsonrpc_request_content(0, method, params)
    response = post(url, payload)
    json_response = response.json()
    return json_response


def read_wallet(file_path):
    """Read keystore file

    :param file_path:

    :return: wallet_info
    """
    if not os.path.isfile(file_path):
        raise FileNotFoundError
    with codecs.open(file_path, 'r', 'utf-8-sig') as f:
        wallet_info = json.load(f)
        f.close()

    return wallet_info


def get_balance_after_transfer(address, uri, request_gen):
    """ Get balance of the address indicated by address for check balance before transfer icx.

    :param address: Icx account address starting with 'hx'
    :param uri: Api uri. type(str)
    :param request_gen:

    :return: Balance of the user's wallet.
    """
    payload_for_balance = get_payload_of_json_rpc_get_balance(address, uri)

    next(request_gen)
    balance_content = request_gen.send(payload_for_balance).json()

    balance = balance_content['result']['response']
    balance_loop = int(balance, 16)
    return balance_loop
